### PR TITLE
added band_as_variable option to open_rasterio

### PR DESCRIPTION
**Pull request by clausmichele**
First draft of a possible implementation of `read_variable_as_bands`.
@snowman2 what do you think? Is it too naive as an approach?

```python
import rioxarray

print(rioxarray.open_rasterio('./results/S2_L2A_data_mod.tiff',band_as_variable=True))
```
```
<xarray.DataArray (band: 3, y: 240, x: 626)>
[450720 values with dtype=uint16]
Coordinates:
  * band         (band) <U3 'B02' 'B03' 'B04'
  * x            (x) float64 6.609e+05 6.609e+05 ... 6.671e+05 6.671e+05
  * y            (y) float64 5.104e+06 5.104e+06 ... 5.102e+06 5.102e+06
    spatial_ref  int64 0
Attributes:
    AREA_OR_POINT:        Area
    PROCESSING_SOFTWARE:  0.6.4a1
    DESCRIPTION:          B02
    _FillValue:           0
    scale_factor:         1.0
    add_offset:           0.0
    long_name:            ('B02', 'B03', 'B04')
```

 - [ ] Closes [#296]
 - [ ] Tests added
 - [ ] Fully documented, including `docs/history.rst` for all changes and `docs/rioxarray.rst` for new API

### CODE REVIEW

**Comment by snowman2** [comment]
Thanks for taking a stab at this @clausmichele :+1: 

The design in #296 would be to load the band data in separate variables so you can do this:
```python
xds = rioxarray.open_rasterio('./results/S2_L2A_data_mod.tiff',band_as_variable=True)
xds["B02"]
```
This enables storing the description, nodata values, etc separately in the attributes of each variable.

An example implementation is here: https://github.com/bopen/xarray-gdal/blob/main/xarray_gdal/xarray_plugin.py

**Comment by clausmichele** [comment]
Ok, second iteration, thanks for the link to `xarray_gdal`, it was really helpful!

image

File: rioxarray/_io.py [hunk 1]
```diff
@@ -1004,36 +1029,43 @@ def open_rasterio(
...
     # always (0, 0, 1) per definition (see
     # https://github.com/sgillies/affine)
-    result.rio.write_transform(riods.transform, inplace=True)
-    rio_crs = riods.crs or result.rio.crs
-    if rio_crs:
-        result.rio.write_crs(rio_crs, inplace=True)
+    # If we have a Dataset with variables and each variable has already 'grid_mapping: spatial_ref'
+    # in the variable attrs, doing the next steps would add another 'spatial_ref' as a
+    # coordinate and will break the functionality of xarray.to_nectdf()
+    # TODO: Probably we would need to check if the transform is already available in each band?
```
Comment (snowman2) [inline rioxarray/_io.py:1061]: The transform should be the same for each band.

File: rioxarray/_io.py [hunk 2]
```diff
@@ -1004,36 +1029,43 @@ def open_rasterio(
...
     # Affine transformation matrix (always available)
     # This describes coefficients mapping pixel coordinates to CRS
     # For serialization store as tuple of 6 floats, the last row being
     # always (0, 0, 1) per definition (see
     # https://github.com/sgillies/affine)
-    result.rio.write_transform(riods.transform, inplace=True)
-    rio_crs = riods.crs or result.rio.crs
-    if rio_crs:
-        result.rio.write_crs(rio_crs, inplace=True)
+    # If we have a Dataset with variables and each variable has already 'grid_mapping: spatial_ref'
```
Comment (snowman2) [inline rioxarray/_io.py:1058]: I recommend not adding the CRS to each variable as you did previously and instead add it to the dataset.

File: rioxarray/_io.py [hunk 3]
```diff
@@ -1050,11 +1082,22 @@ def open_rasterio(
...
-    return result
+    # remove duplicate tags for DataArray
+    if not output_is_dataset: # result.name is not available in a Dataset
+        if result.name:
+            result.attrs = {
+                attr: value
+                for attr, value in result.attrs.items()
+                if not attr.startswith(f"{result.name}#")
+            }
+    # remove duplicate tags for Dataset, otherwise there are issues opening the netCDF file in QGIS
```
Comment (snowman2) [inline rioxarray/_io.py:1093]: This is interesting. Mind adding a separate function called `_clean_attrs` and moving the duplicate cleanup there?

File: rioxarray/_io.py [hunk 4]
```diff
@@ -967,7 +974,26 @@ def open_rasterio(
...
     else:
         coord_name = "band"
-        coords[coord_name] = np.asarray(riods.indexes)
+        band_descriptions = riods.descriptions
+        # Assign the band names only if the description is available for all of them
+        if band_as_variable and not any(map(lambda ele: ele is None, band_descriptions)):
+            output_is_dataset = True
+            data_vars = {}
+            for i in riods.indexes:
+                data_var_attrs = {
```
Comment (snowman2) [inline rioxarray/_io.py:983]: I am wondering if band specific tags should be added here?

**Comment by snowman2** [comment]
@clausmichele thanks for working on this. I added some comments and will try to do a more thorough review later.

**Comment by snowman2** [comment]
Okay, this got me down a rabbit hole to a different implementation to preserve the lazy loading (see #600). Want to give it a try and see what you think?

**Pull request closed without merging**